src/sources/postgresql_metrics.rs

```diff
@@ -0,0 +1,840 @@
+use crate::{
+    config::{DataType, GlobalOptions, SourceConfig, SourceDescription},
+    event::metric::{Metric, MetricKind, MetricValue},
+    internal_events::{PostgresqlMetricsCollectCompleted, PostgresqlMetricsCollectFailed},
+    shutdown::ShutdownSignal,
+    Event, Pipeline,
+};
+use chrono::{DateTime, Utc};
+use futures::{
+    future::{join_all, try_join_all},
+    stream, FutureExt, SinkExt, StreamExt, TryFutureExt,
+};
+use openssl::{
+    error::ErrorStack,
+    ssl::{SslConnector, SslMethod},
+};
+use postgres_openssl::MakeTlsConnector;
+use serde::{Deserialize, Serialize};
+use snafu::{ResultExt, Snafu};
+use std::{
+    collections::{BTreeMap, HashSet},
+    future::ready,
+    path::PathBuf,
+    time::Instant,
+};
+use tokio::time;
+use tokio_postgres::{
+    config::{ChannelBinding, Host, SslMode, TargetSessionAttrs},
+    types::FromSql,
+    Client, Config, Error as PgError, NoTls, Row,
+};
+
+macro_rules! tags {
+    ($tags:expr) => { $tags.clone() };
+    ($tags:expr, $($key:expr => $value:expr),*) => {
+        {
+            let mut tags = $tags.clone();
+            $(
+                tags.insert($key.into(), $value.into());
+            )*
+            tags
+        }
+    };
+}
+
+macro_rules! counter {
+    ($value:expr) => {
+        MetricValue::Counter {
+            value: $value as f64,
+        }
+    };
+}
```

Comment:
Should we extract these into a shared utility module?

Reply:
We have an issue for this -- https://github.com/timberio/vector/issues/5229